src/cfg/domtree.h

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2021 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+//
+// Computes a dominator tree for a CFG graph, that is, for each block x we find
+// the blocks y_i that must be reached before x on any path from the entry. Each
+// block x has a single immediate dominator, the one closest to it, which forms
+// a tree structure.
+//
+
+#ifndef domtree_h
+#define domtree_h
+
+#include "cfg-traversal.h"
+#include "wasm.h"
+
+namespace wasm {
+
+//
+// DomTree receives an input CFG which has a list of basic blocks in reverse
+// postorder. It generates the dominator tree by representing it as a vector of
+// indexes, for each block giving the index of its parent (the immediate
+// dominator) in the tree, that is,
+//
+//  parents[0] = a nonsense value, as the entry node has no immediate dominator
+//  parents[1] = the index of the immediate dominator of CFG.blocks[1]
+//  parents[2] = the index of the immediate dominator of CFG.blocks[2]
+//  etc.
+//
+// The BasicBlock type is assumed to have a ".in" property which declares a
+// vector of pointers to the incoming blocks, that is, the predecessors.
+template<typename BasicBlock> struct DomTree {
```

Comment:
Why is this a template class? Do you plan to use this on other than `BasicBlock`? Oh, I noticed you defined another `BasicBlock` class in test/domtree.cpp, but is that the reason? That test file can include a Binaryen header that defines `BasicBlock` I guess, and if this is not a template class, we can move the algorithm to a cpp file.

Reply:
This is because CFGWalker is a template itself, and the BasicBlock type inside it depends on the template parameter. So there isn't a single type here.

There might be a clever C++ way to solve this, like moving BasicBlock out of CFGWalker and subclassing it inside to add the data with the templated type, but I can't see a simple way to do that which works...

Reply:
Why should `BasicBlock` be inside `CFGWalker`?

Reply:
Currently CFGWalker creates and owns a list of BasicBlocks. Subclasses of CFGWalker add more fields to BasicBlock as needed. Maybe this could all be restructured, but it would probably need more dynamic checks I think (right now it uses CRTP).